tealish/nodes.py

```diff
@@ -88,91 +109,98 @@ def new_scope(self, name="", slot_range=None):
             else name,
         }
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         name = self.__class__.__name__
         return name
 
 
 class Expression(Node):
     @classmethod
-    def parse(cls, string, parent, compiler):
-        return cls(string)
+    def parse(cls, line: str, parent: Node, compiler: "TealishCompiler") -> Node:
+        return cls(line)
 
     @classmethod
-    def match(cls, string):
-        return re.match(cls.pattern, string) is not None
+    def match(cls, line: str) -> bool:
+        return re.match(cls.pattern, line) is not None
 
 
 class Literal(Expression):
     pattern = rf"(?P<value>{LITERAL_BYTES}|{LITERAL_INT})$"
 
     @classmethod
-    def parse(cls, string, parent, compiler):
-        for expr in [LiteralInt, LiteralBytes]:
-            if expr.match(string):
-                return expr.parse(string, parent, compiler)
-        raise ParseError(f'Cannot parse "{string}" as Literal')
+    def parse(cls, line: str, parent: Node, compiler: "TealishCompiler") -> Node:
+        matchable: List[Type[Expression]] = [LiteralInt, LiteralBytes]
+        for expr in matchable:
+            if expr.match(line):
+                return expr.parse(line, parent, compiler)
+        raise ParseError(f'Cannot parse "{line}" as Literal')
 
 
 class LiteralInt(Expression):
     pattern = rf"(?P<value>{LITERAL_INT})$"
     value: int
 
-    def write_teal(self, writer):
+    def write_teal(self, writer: "TealWriter"):
         writer.write(self, f"pushint {self.value}")
 
-    def type(self):
+    def type(self) -> str:
         return "int"
 
-    def _tealish(self, formatter=None):
+    def _tealish(self, formatter=None) -> str:
         return f"{self.value}"
 
 
 class LiteralBytes(Expression):
     pattern = rf"(?P<value>{LITERAL_BYTES})$"
     value: str
 
-    def write_teal(self, writer):
+    def write_teal(self, writer: "TealWriter"):
         writer.write(self, f"pushbytes {self.value}")
 
-    def type(self):
+    def type(self) -> str:
         return "bytes"
 
-    def _tealish(self, formatter=None):
+    def _tealish(self, formatter=None) -> str:
         return f"{self.value}"
 
 
 class Name(Expression):
     pattern = rf"(?P<value>{VARIABLE_NAME})$"
     value: str
 
-    def __init__(self, string) -> None:
-        self.slot = None
-        self._type = None
-        super().__init__(string)
+    def __init__(self, line: str):
+        self.slot: Optional[int] = None
+        self._type: Optional[str] = None
+        super().__init__(line)
 
-    def _tealish(self, formatter=None):
+    def _tealish(self, formatter=None) -> str:
         return f"{self.value}"
 
-    def type(self):
+    def type(self) -> Optional[str]:
         return self._type
 
 
 class GenericExpression(Expression):
+
+    # TODO: never set?
+    type: str
```

Comment:
Added only to appease mypy, idk if its ever set on a GenericExpression

Reply:
It doesn't get set. `GenericExpression.parse` actually returns a `Node` rather than an instance of `GenericExpression`

Reply:
The reason this is added is that some attributes on nodes are typed as a GenericExpression, so maybe those should be changed to the actual Node type